File: backend/internal/apps/speechclip/timeline.go
package speechclip

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
)

// Timeline is the v1 timelines/main.json contract for the Content Studio
// Remotion timeline preview. Go validation here is authoritative; the TS
// counterpart in frontend/src/apps/speechclip/timeline.ts is UI-only early
// feedback.
//
// Field alignment with JSONL sources:
//   - Clip.SourceSentenceIDs → transcript `i` field
//   - Clip.StartMs / EndMs   → transcript `start` / `end` (already in ms)
//   - Clip.Text              → highlight `corrected_text` (or raw `text`)
type Timeline struct {
	Version    int            `json:"version"`              // must be 1
	ID         string         `json:"id"`                   // must be "main"
	AssetID    string         `json:"assetId"`              // root asset id, non-empty
	DurationMs *int           `json:"durationMs,omitempty"` // optional; must be >= last clip endMs
	Clips      []TimelineClip `json:"clips"`                // non-empty ordered edit segments
}

// TimelineClip is one edit segment. SourceSentenceIDs references the `i` field
// from transcripts/<assetId>.jsonl; Text is typically corrected_text from
// highlights/<assetId>.jsonl.
type TimelineClip struct {
	StartMs           int    `json:"startMs"`           // ms, 0 <= startMs < endMs
	EndMs             int    `json:"endMs"`             // ms
	Text              string `json:"text"`              // display text (corrected_text or raw)
	SourceSentenceIDs []int  `json:"sourceSentenceIds"` // sentence `i` values this clip covers
}

// ValidateTimeline enforces the authoritative v1 contract constraints.
func ValidateTimeline(t *Timeline) error {
	if t.Version != 1 {
		return fmt.Errorf("version must be 1, got %d", t.Version)
	}
	if t.ID != "main" {
		return fmt.Errorf("id must be \"main\", got %q", t.ID)
	}
	if t.AssetID == "" {
		return errors.New("assetId is required")
	}
	if len(t.Clips) == 0 {
		return errors.New("clips must be non-empty")
	}
	for idx, c := range t.Clips {
		if c.StartMs < 0 {
			return fmt.Errorf("clip[%d]: startMs must be >= 0, got %d", idx, c.StartMs)
		}
		if c.StartMs >= c.EndMs {
			return fmt.Errorf("clip[%d]: startMs (%d) must be < endMs (%d)", idx, c.StartMs, c.EndMs)
		}
	}
	if t.DurationMs != nil {
		d := *t.DurationMs
		if d <= 0 {
			return fmt.Errorf("durationMs must be > 0, got %d", d)
		}
		last := t.Clips[len(t.Clips)-1]
		if d < last.EndMs {
			return fmt.Errorf("durationMs (%d) must be >= last clip endMs (%d)", d, last.EndMs)
		}
	}
	return nil
}

// loadTimeline reads and validates timelines/main.json for a workspace.
func loadTimeline(ws string) (*Timeline, error) {
	data, err := os.ReadFile(filepath.Join(appDir(ws), "timelines", "main.json"))
	if err != nil {
		return nil, err
	}
	var t Timeline
	if err := json.Unmarshal(data, &t); err != nil {
		return nil, fmt.Errorf("parse error: %w", err)
	}
	if err := ValidateTimeline(&t); err != nil {
		return nil, fmt.Errorf("invalid timeline: %w", err)
	}
	return &t, nil
}

// saveTimeline validates and writes timelines/main.json.
func saveTimeline(ws string, t *Timeline) error {
	if err := ValidateTimeline(t); err != nil {
		return fmt.Errorf("invalid timeline: %w", err)
	}
	dir := filepath.Join(appDir(ws), "timelines")
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	data, _ := json.MarshalIndent(t, "", "  ")
	return os.WriteFile(filepath.Join(dir, "main.json"), data, 0o644)
}

File: frontend/src/apps/speechclip/timeline.ts
/**
 * v1 timelines/main.json contract — TypeScript types and UI-side validation.
 *
 * Go backend (backend/internal/apps/speechclip/timeline.go) is the authoritative
 * validator on save. This module provides early feedback in the UI so users see
 * errors before the round-trip.
 *
 * Field alignment with JSONL sources:
 *   SourceSentenceIDs → transcript `i`
 *   startMs / endMs   → transcript `start` / `end` (already in ms)
 *   text              → highlight `corrected_text` (or raw `text`)
 */

export interface TimelineClip {
    /** ms, 0 <= startMs < endMs; maps to transcript `start` */
    startMs: number;
    /** ms; maps to transcript `end` */
    endMs: number;
    /** display text — typically highlight `corrected_text` or raw `text` */
    text: string;
    /** sentence `i` values from transcripts/<assetId>.jsonl that this clip covers */
    sourceSentenceIds: number[];
}

export interface Timeline {
    /** must be 1 */
    version: 1;
    /** must be "main" */
    id: 'main';
    /** root asset id, non-empty */
    assetId: string;
    /** optional total duration in ms; must be >= last clip endMs */
    durationMs?: number;
    /** ordered edit segments, non-empty */
    clips: TimelineClip[];
}

/**
 * validateTimeline returns an error message string, or null if valid.
 * Mirrors ValidateTimeline in timeline.go — keep the two in sync.
 */
export function validateTimeline(t: unknown): string | null {
    if (typeof t !== 'object' || t === null) return 'timeline must be an object';
    const tl = t as Record<string, unknown>;

    if (tl['version'] !== 1) return `version must be 1, got ${tl['version']}`;
    if (tl['id'] !== 'main') return `id must be "main", got ${JSON.stringify(tl['id'])}`;
    if (!tl['assetId'] || typeof tl['assetId'] !== 'string') return 'assetId is required';

    const clips = tl['clips'];
    if (!Array.isArray(clips) || clips.length === 0) return 'clips must be non-empty';

    for (let idx = 0; idx < clips.length; idx++) {
        const c = clips[idx] as Record<string, unknown>;
        const startMs = c['startMs'];
        const endMs = c['endMs'];
        if (typeof startMs !== 'number' || startMs < 0) {
            return `clip[${idx}]: startMs must be >= 0, got ${startMs}`;
        }
        if (typeof endMs !== 'number' || startMs >= endMs) {
            return `clip[${idx}]: startMs (${startMs}) must be < endMs (${endMs})`;
        }
    }

    if (tl['durationMs'] !== undefined) {
        const d = tl['durationMs'];
        if (typeof d !== 'number' || d <= 0) return `durationMs must be > 0, got ${d}`;
        const lastClip = clips[clips.length - 1] as Record<string, unknown>;
        const lastEnd = lastClip['endMs'] as number;
        if (d < lastEnd) return `durationMs (${d}) must be >= last clip endMs (${lastEnd})`;
    }

    return null;
}

// ── fixtures ────────────────────────────────────────────────────────────────

/**
 * Valid timeline fixture — derived from the smoke-test transcript sentences:
 *   i=0 start=0    end=3000  corrected_text="今天是一个硬件的采购记录。"
 *   i=2 start=6000 end=9000  corrected_text="芯片是ESP32，不支持经典蓝牙。"
 */
export const VALID_TIMELINE_FIXTURE: Timeline = {
    version: 1,
    id: 'main',
    assetId: 'a01',
    clips: [
        {
            startMs: 0,
            endMs: 3000,
            text: '今天是一个硬件的采购记录。',
            sourceSentenceIds: [0],
        },
        {
            startMs: 6000,
            endMs: 9000,
            text: '芯片是ESP32，不支持经典蓝牙。',
            sourceSentenceIds: [2],
        },
    ],
};

/**
 * Invalid-timestamp fixture — startMs > endMs, should fail validation.
 * Used in tests to verify the validator catches reversed timestamps.
 */
export const INVALID_TIMESTAMP_FIXTURE: Record<string, unknown> = {
    version: 1,
    id: 'main',
    assetId: 'a01',
    clips: [{ startMs: 5000, endMs: 3000, text: 'reversed', sourceSentenceIds: [0] }],
};
